### ISSUE TITLE
`pipenv lock` fails when executed under python 3.12

### ISSUE THREAD

**Issue by rvodden** [pypa/pipenv#6010]
Be sure to check the existing issues (both open and closed!), and make sure you are running the latest version of Pipenv.

Check the [diagnose documentation](https://pipenv.pypa.io/en/latest/diagnose/) for common issues before posting! We may close your issue if it is very similar to one of them. Please be considerate, or be on your way.

Make sure to mention your debugging experience if the documented solution failed.

### Issue description

Runing pipenv lock on python 3.12 causes pipenv to fail with:

```
  File "/tmp/pip-build-env-cxa535v1/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 10, in <module>
    import distutils.core
ModuleNotFoundError: No module named 'distutils'
```

### Expected result

`pipenv lock` would create or update a `Pipenv.lock` file

### Actual result

When possible, provide the verbose output (`--verbose`), especially for locking and dependencies resolving issues.

```
[rvodden:~/Source/maaspower]└2 [maaspower] fix_linting+* 4s 1 ± pipenv lock --verbose
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
INFO:pipenv.patched.pip._internal.operations.prepare:Obtaining file:///home/rvodden/Source/maaspower (from -r 
/tmp/pipenv-wnki6z45-requirements/pipenv-r79h4iyg-constraints.txt (line 2))
INFO:pipenv.patched.pip._internal.cli.spinners:Installing build dependencies: started
INFO:pipenv.patched.pip._internal.cli.spinners:Installing build dependencies: finished with status 'done'
INFO:pipenv.patched.pip._internal.cli.spinners:Checking if build backend supports build_editable: started
INFO:pipenv.patched.pip._internal.cli.spinners:Checking if build backend supports build_editable: finished with status 'done'
Traceback (most recent call last):
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/resolver.py", line 675, in <module>
    main()
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/resolver.py", line 661, in main
    _main(
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/resolver.py", line 645, in _main
    resolve_packages(
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/resolver.py", line 612, in resolve_packages
    results, resolver = resolve(
                        ^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/resolver.py", line 592, in resolve
    return resolve_deps(
           ^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/utils/resolver.py", line 908, in resolve_deps
    results, hashes, internal_resolver = actually_resolve_deps(
                                         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/utils/resolver.py", line 681, in 
actually_resolve_deps
    resolver.resolve()
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/utils/resolver.py", line 440, in resolve
    results = resolver.resolve(constraints, check_supported_wheels=False)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/resolver.py"
, line 76, in resolve
    collected = self.factory.collect_root_requirements(root_reqs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/factory.py",
line 513, in collect_root_requirements
    reqs = list(
           ^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/factory.py",
line 474, in _make_requirements_from_install_req
    cand = self._make_candidate_from_link(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/factory.py",
line 190, in _make_candidate_from_link
    self._editable_candidate_cache = EditableCandidate(
                                           ^^^^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.p
y", line 322, in __init__
    super().__init__(
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.p
y", line 157, in __init__
    self.dist = self._prepare()
                ^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.p
y", line 226, in _prepare
    dist = self._prepare_distribution()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.p
y", line 332, in _prepare_distribution
    return self._factory.preparer.prepare_editable_requirement(self._ireq)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/operations/prepare.py", 
line 696, in prepare_editable_requirement
    dist = _get_prepared_distribution(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/operations/prepare.py", 
line 71, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/distributions/sdist.py", 
line 52, in prepare_distribution_metadata
    self.req.isolated_editable_sanity_check()
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/req/req_install.py", line 
544, in isolated_editable_sanity_check
    and not self.supports_pyproject_editable()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_internal/req/req_install.py", line 
247, in supports_pyproject_editable
    return "build_editable" in self.pep517_backend._supported_features()
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_vendor/pyproject_hooks/_impl.py", 
line 153, in _supported_features
    return self._call_hook('_supported_features', {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_vendor/pyproject_hooks/_impl.py", 
line 321, in _call_hook
    raise BackendUnavailable(data.get('traceback', ''))
pipenv.patched.pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):
  File 
"/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/patched/pip/_vendor/pyproject_hooks/_in_process/_in_proc
ess.py", line 77, in _build_backend
    obj = import_module(mod_path)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name, package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1304, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1325, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 929, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 994, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/tmp/pip-build-env-fgmj84tv/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 10, in <module>
    import distutils.core
ModuleNotFoundError: No module named 'distutils'
✘ Locking Failed!
⠙ Locking...
Traceback (most recent call last):
  File "/home/rvodden/.pyenv/versions/maaspower/bin/pipenv", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/cli/options.py", line 58, in main
    return super().main(*args, **kwargs, windows_expand_args=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/decorators.py", line 92, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/vendor/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/cli/command.py", line 340, in lock
    do_lock(
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/routines/lock.py", line 65, in do_lock
    venv_resolve_deps(
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/utils/resolver.py", line 849, in venv_resolve_deps
    c = resolve(cmd, st, project=project)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rvodden/.pyenv/versions/3.12.0/envs/maaspower/lib/python3.12/site-packages/pipenv/utils/resolver.py", line 718, in resolve
    raise RuntimeError("Failed to lock Pipfile.lock!")
RuntimeError: Failed to lock Pipfile.lock!
```

### Steps to replicate

Provide the steps to replicate (which usually at least includes the commands and the Pipfile).

1. install python 3.12
2. set python 3.12 to be your current python
3. install pyenv
4. create python project
5. run pipenv lock

-------------------------------------------------------------------------------

Please run `$ pipenv --support`, and paste the results here. Don't put backticks (`` ` ``) around it! The output already contains Markdown formatting.

If you're on macOS, run the following:

    $ pipenv --support | pbcopy

If you're on Windows, run the following:

    > pipenv --support | clip

If you're on Linux, run the following:

    $ pipenv --support | xclip

pipenv support is here:

https://gist.github.com/rvodden/7c39c1984c36d1f45f788bbf5ec91ee3

**Comment by matteius** [comment]
The test runner runs on python 3.12 and there are plenty of operations that lock.

**Comment by rvodden** [comment]
Ok. I can still replicate this reliably by using the steps I gave in the report.

**Comment by Farit** [comment]
hey @rvodden 

I had the same issue. This error has nothing to do with the Pipenv itself. The problem lies in a particular library specified in your Pipfile. The [newly released Python 3.12 removed distutils](https://docs.python.org/3/whatsnew/3.12.html) after it has been deprecated since Python 3.10. You need to upgrade some libraries to support Python 3.12.
